### PR TITLE
network: support interface binding on Android

### PR DESCRIPTION
**Pull request by goaway**
Description: `SO_BINDTODEVICE` will generally only be useable by an Android app running as root, so the present interface-binding approach on Android is mostly a non-starter. Instead, this change binds the interface indirectly by finding the local IP of the interface and setting that as the source IP of the socket.

To do so, it passes a "synthetic" socket option through to the internal connection API. When applied, instead of setting an actual socket option on the socket, this object sets the source IP property on the connection before bind() is called.
Risk Level: Moderate
Testing: Local & On Device

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

File: library/common/network/src_addr_socket_option_impl.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include "envoy/network/listen_socket.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * This is a "synthetic" socket option implementation, which sets the source IP/port of a socket
+ * using a provided IP address (and maybe port) during bind.
+ *
+ * Based on the OriginalSrcSocketOption extension.
```
[the remark this answers is not in the document]
Comment (snowp) [reply library/common/network/src_addr_socket_option_impl.h:14]: I think perhaps ideally we'd lift this up to a generic socket option (so it's not in extension code) and have EM tests that it works the right way for EM's use case so we catch any regression

That said I'm not gonna block on this, if you think it's easier to copy this I would just do that (and maybe note that at the time of writing it's the same as the OriginalSrc one)

[the remark this answers is not in the document]
Comment (goaway) [reply library/common/network/src_addr_socket_option_impl.h:14]: What still gives me pause is that I'm not sure this is so much a case of shared inherent behavior, so much as shared incidental behavior.

My preference right now is to move forward, but I'm open to continuing the discussion.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Comment by goaway** [comment]
Thanks @snowp!

**Pull request merged**